backend/src/org/akvo/dash/transformation.clj

```diff
@@ -1,225 +1,157 @@
-;;   "Update column type to new-type on column with columns-title."
-;;   [data column-title new-type]
-;;   (-> data
-;;       (update (*> (in [:columns])
-;;                   each
-;;                   (conditionally #(= column-title (get % :title))))
-;;               #(assoc % column-title column-type))))
-
-;; (update-column-type
-;;  {:columns [{:title "Name" :type "STRING"} {:title "Age" :type "STRING"}]}
-;;  "Age"
-;;  "NUMBER"
-;;  )
-
-
-;; (defn coerce-column-view
-;;   ""
-;;   [data column-title type]
-;;   (-> data
-;;      (view (*> (in [:columns])
-;;                each
-;;                (conditionally #(= column-title (get % :title)))
-;;                (in [:values])))))
-
-
-;; (defn coerce-column
-;;   ""
-;;   [data column-title type]
-;;   (-> data
-;;       (update (*> (in [:columns])
-;;                 each
-;;                 (conditionally #(= column-title (get % :title)))
-;;                 (in [:values]))
-;;               #(map read-string %))))
-
-;; (pprint
-;;  (coerce-column
-;;   {:columns [{:title "Name" :type "STRING" :values ["1" "2" "3"]}
-;;              {:title "Age" :type "STRING" :values ["1" "2" "3"]}]}
-;;   "Age"
-;;   "NUMBER"))
-
-
-
-
-;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
-;; Playing with traversy lenses
+  (:require [clojure.java.jdbc :as jdbc]
+            [clojure.string :as str]
+            [hugsql.core :as hugsql]
+            [org.akvo.dash.transformation.engine :as engine]
+            [org.akvo.dash.util :refer (squuid gen-table-name)]))
+
+;; TODO: Potential change op-spec validation `core.spec`
+
+(def ops-set (set (keys engine/available-ops)))
+(def on-error-set #{"fail" "default-value" "delete-row"})
+(def type-set #{"number" "text" "date"})
+(def sort-direction-set #{"ASC" "DESC"})
+
+(hugsql/def-db-fns "org/akvo/dash/transformation.sql")
+
+
+(defn- throw-invalid-op
+  [op-spec]
+  (throw (Exception. (str "Invalid operation " op-spec))))
+
+(defn required-keys
+  [{:strs [op args onError] :as op-spec}]
+  (or (boolean (and (ops-set op)
+                    (not-empty args)
+                    (re-find #"c\d+" (args "columnName"))
+                    (on-error-set onError)))
+      (throw-invalid-op op-spec)))
+
+(defmulti valid-op?
+  (fn [op-spec]
+    (keyword (op-spec "op"))))
+
+(defmethod valid-op? :default
+  [op-spec]
+  (throw-invalid-op op-spec))
+
+(defmethod valid-op? :core/change-datatype
+  [{:strs [op args] :as op-spec}]
+  (or (boolean (and (required-keys op-spec)
+                 (type-set (args "newType"))
+                 (not-empty (args "defaultValue"))
+                 (if (= "date" (args "newType"))
+                   (not-empty (args "parseFormat"))
+                   true)))
+      (throw-invalid-op op-spec)))
+
+(defmethod valid-op? :core/change-column-title
+  [{:strs [args] :as op-spec}]
+  (or (boolean (and (required-keys op-spec)
+                    (not-empty (args "columnTitle"))))
```

Comment:
`(not-empty (args "columnTitle"))` is roughly the same as `(contains? args "columnTitle")`.